pallets/crowdloans/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn vaults)]
+    pub type Vaults<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        ParaIdOf,
+        Vault<ParaId, AssetIdOf<T>, BalanceOf<T>>,
+        OptionQuery,
+    >;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        ////
+        //// 1. Vaults Management
+
+        /// - `token` is the currency or token which needs to be deposited to fill
+        ///   the vault and later participate in the crowdloans
+        /// - `crowdloan` represents which crowdloan we are supporting
+        /// - `project_shares` and `currency_shares` are new assets created for this
+        ///   vault to represent the shares of the vault's contributors
+        /// - `until` sets the vault's expiration block, a vault must be "expired"
+        ///   until it can be used to participate in its crowdloan
+        #[pallet::weight(10_000)]
+        #[allow(unused)]
+        pub fn create_vault(
+            origin: OriginFor<T>,
+            token: AssetIdOf<T>,
+            crowdloan: ParaId,
+            project_shares: AssetIdOf<T>,
+            currency_shares: AssetIdOf<T>,
+            contribution_strategy: ContributionStrategy<ParaId, AssetIdOf<T>, BalanceOf<T>>,
+            claim_strategy: ClaimStrategy<ParaId>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            // get
+            let project_shares_issuance = T::CrowdloanCurrency::total_issuance(project_shares);
+            let currency_shares_issuance = T::CrowdloanCurrency::total_issuance(currency_shares);
+
+            // make sure both project_shares and currency_shares are new assets
+            ensure!(
+                project_shares_issuance == Zero::zero() && currency_shares_issuance == Zero::zero(),
+                Error::<T>::SharesNotNew
+            );
+
+            // make sure no similar vault already exists as identified by crowdloan
+            ensure!(
+                !Vaults::<T>::contains_key(&crowdloan),
+                Error::<T>::CrowdloanAlreadyExists
+            );
+
+            // add new vault to vaults storage
+            Vaults::<T>::try_mutate(&crowdloan, |vault| -> Result<_, DispatchError> {
+                // inialize new vault
+                Ok(*vault = Some(crowdloan_structs::Vault {
+                    project_shares: project_shares,
+                    currency_shares: currency_shares,
+                    currency: currency_shares,
+                    phase: VaultPhase::CollectingContributions,
+                    contribution_strategy: ContributionStrategy::Placeholder(
+                        crowdloan,
+                        currency_shares,
+                        0,
+                    ),
+                    claim_strategy: ClaimStrategy::Placeholder(crowdloan),
+                    contributed: 0,
+                }))
+            })
+        }
```

Comment:
We should make sure no vault for the same parachain already exists

Reply:
@ETeissonniere i'm not sure I understand how the check above https://github.com/parallel-finance/parallel/blob/14ebaade49171edf537329c432413f751aa7377b/pallets/crowdloans/src/lib.rs#L144-L148 is insufficient? Where else should be checked before mutating? Thanks!